### ISSUE TITLE
DiD: allow for multiple pre and post intervention observations

### ISSUE THREAD

**Issue by drbenvincent** [pymc-labs/CausalPy#76]
At the moment the code works but it assumes there are observations at _one_ pre and _one_ post intervention time.

So we need to relax this assumption and generalise the code.

Probably best to do this while working through the 'bank failure' dataset, see #44.

This will also impact plotting

And it will impact reporting of the summery stats.

## Bank failure dataset/example + robustifying
- [x] add data
- [x] data processing + visualisation
- [x] need to add `treated` and `units` columns
- [x] check if I need to either specify the grouping column (e.g. `district`) as categorical OR enter that as `C(district)` in the model formula

## Classic 2x2 DID
- [x] Analysis 1 with just pre and post (Bayesian)
  - [x] model fit working
  - [x] plotting working
- [ ] ~~Implement example for the frequentist approach~~

## 'Extended' DID with more than 2 observed time points
- [x] Analysis 2 with all data
  - [ ] Consider if we need to have separate classes for classic 2x2 DID and another for more time points, or try to make the one `DifferenceInDifferences` class deal with all situations.
  - [x] model fit working
  - [ ] plotting working
- [ ] ~~Make it work for the frequentist model~~
- [ ] Update tests
- [ ] Update `README` + `index.rst`

**Comment by drbenvincent** [comment]
Model formula will be `bib ~ 1 + district + year + district:treated`

**Comment by drbenvincent** [comment]
At this point it seems that model fitting _does_ work, although we'll have to wait for the visualisation to see if the results make sense.

Plotting is going to get more complex. The initial plot that I had only makes sense when we have two time points (pre/post) AND we have multiple observed units per group.

So we've potentially got a 2*2 grid of different plot types to produce

|                             | pre/post only | multiple time points |
|-----------------------------|---------------|----------------------|
| one unit observed per group |               |                      |
| multiple units per group    | existing plot |                      |

**Comment by drbenvincent** [comment]
![Screenshot 2022-11-19 at 21 31 32](https://user-images.githubusercontent.com/6765047/202872350-1512bb16-9115-404b-b6dc-4589e53fd30b.png)

- [ ] fix green arrow for causal impact
- [ ] fix plotting of data

**Comment by drbenvincent** [comment]
The magnitude of the causal impact is wrong. I think this might be fixed by enforcing an order on the levels of the groups.

**Comment by drbenvincent** [comment]
Plotting now works for the original dataset (multiple units in the treatment and control conditions) and the new banking example (one unit in the treatment and one in the control condition). Although the aesthetics could do with some work, the priority is to focus on getting it working in the case where we have more observations over time, not just the pre/post times.

Original dataset
![Screenshot 2022-11-20 at 12 31 55](https://user-images.githubusercontent.com/6765047/202902021-3622b1a7-686e-4105-a0eb-9490473494d0.png)

Banking dataset
![Screenshot 2022-11-20 at 12 32 17](https://user-images.githubusercontent.com/6765047/202902042-17bc30d4-ad6f-45a5-a171-4c5ab578affa.png)

**Comment by drbenvincent** [comment]
Currently sampling from the posterior works with 

```python
result = DifferenceInDifferences(
    df_long,
    formula="bib ~ 1 + district + year + district:treated",
    time_variable_name="year",
    group_variable_name="district",
    treated="Sixth District",
    untreated="Eighth District",
    prediction_model=LinearRegression()
    )
```

But breaks when we get to doing the other stuff and expected. This is the next step.

**Comment by drbenvincent** [comment]
Currently working in the `did_multiple_observations` branch.

**Comment by drbenvincent** [comment]
At this point we have DiD working for:
1. Single pre and post treatment observation (although we have some shape issues, calculating posterior predictions and counterfactuals for each item)
<img width="606" alt="Screenshot 2022-12-25 at 21 29 46" src="https://user-images.githubusercontent.com/6765047/209482365-9c8cca64-0d11-48f2-82f5-4184999e6163.png">

But there are clearly some issues to be resolved for the `banks` dataset.
<img width="605" alt="Screenshot 2022-12-25 at 21 30 39" src="https://user-images.githubusercontent.com/6765047/209482400-3ff36561-10c2-4718-9ed7-2ccd1ec2db41.png">

2. Multiple pre and post treatment observations.

Looks like this for the full `banks` dataset

<img width="604" alt="Screenshot 2022-12-25 at 21 31 28" src="https://user-images.githubusercontent.com/6765047/209482412-f33cc500-531d-4af5-b50d-12125d519f4c.png">

**Comment by drbenvincent** [comment]
In the `banks` dataset, we are getting 1 surplus degree of freedom. I think it is because the group is coded as a string (therefore treated as a category) rather than as a numerical 0/1.

**Comment by drbenvincent** [comment]
I've made meaningful improvements to DiD at this point. There were a number of things about the code before which were muddled and a bit wrong. I've fixed those up, made the code cleaner, got DiD working for multiple pre and post treatment observations, and improved the plotting.

![notebooks_did_pymc_banks_12_0](https://user-images.githubusercontent.com/6765047/209569647-0a038eb1-abe4-479f-8352-ac45b0c91474.png)
![notebooks_did_pymc_banks_17_0](https://user-images.githubusercontent.com/6765047/209569648-737b1672-5c32-4005-8ebb-c68a9d4191d9.png)
